### PR TITLE
Authority

### PR DESCRIPTION
**Pull request by TomShacham**
addresses #3 

not entirely sure about the regex i've used: "(?:([^:]+)(?:\:([^@]*)@))?([^:]+)(?:\:(\\d+))?"
the RFC regex ("(?:([^@]+)@)?([^:]+)(?:\:([\d]+))?") does not seem to do what i expect.

tests i've written should cover all the cases so would love a 2nd pair of eyes 👀

### CODE REVIEW

**Comment by TomShacham** [comment]
also i've duplicated our own regex class in this PR, we can tidy up if/when this PR and the PR addressing #2 is merged.

File: contract/src/uri.ts [hunk 1]
```diff
@@ -38,9 +39,8 @@ export class Uri implements ParsedUri {
   /** {@link https://tools.ietf.org/html/rfc3986#section-5.3} */
   toString() {
     const result: string[] = [];
-
     if (typeof this.scheme != 'undefined') result.push(this.scheme, ":");
-    if (typeof this.authority != 'undefined') result.push("//", this.authority);
+    if (typeof this.authority != 'undefined' && typeof this.authority.host != 'undefined') result.push("//", this.authority.toString());
```
Comment (danielbodart) [inline contract/src/uri.ts]: You shouldn't need to do this: typeof this.authority.host != 'undefined'

Your breaking encapsulation

**Comment by danielbodart** [comment]
Does the regex come from the RFC or did you pull it from utterlyidle? 

if it's from utterlyidle then I don't think I ever bothered with userinfo because they deprecated it in http (for security reasons) but i think it's good you have as it's useful

**Comment by savagematt** [comment]
Side-stepped the PR and copy-pasted the code because I was too lazy to rebase from master.

https://github.com/http4t/http4t/commit/c78c84741aa2013942863fc36f609f5af3f19b3a

**Pull request closed without merging**